<style>
body {
    max-width: 900px;
    margin: 40px auto;
    padding: 0 20px;
    font-family: "Georgia", serif;
    line-height: 1.6;
}
</style>

<div style="text-align: center; padding: 60px 60px">
  <h1 style="font-weight: bold; font-size: 3.1em">
    ENCONTRAR EL MEJOR ARIMA PARA LA SERIE DIFERENCIADA
  </h1>
</div>

Esta vez encontraremos un modelo variando la parte media móvil estacional y la parte autorregresiva estacional.

**RESULTADOS**:
- No hay modelos propuestos. Todos tienen dependencia en los residuos.
- `aic_values-4.csv`
- `AIC_3D_plot-4.html`.

# **CARGAR LOS DATOS**

In [1]:
from statsmodels.tsa.statespace.sarimax import SARIMAX
import pandas as pd

data=pd.read_csv('MXN00021035.csv')

pre=data.iloc[:,6]  # Precipitacion, es la columna 5
date=data.iloc[:,5] # Date, es la columna 6
date = date.astype(str).str.replace(r'(\d{4})(\d{2})', r'\1/\2', regex=True)    # La fecha está como 195210 y la pasamos a 1952/10 
date = pd.to_datetime(date, format='%Y/%m')                                     # Lo convertimos en fecha
pre = pd.Series(pre.values, index=date)                                         # Creamos una Serie

# Partir la serie para train y test
pre_total = pre.copy()          # Copia de la serie original

# Todas hasta los ultimos 12 meses
pre = pre_total[:-12]           # Entrenamiento: todos menos los últimos 12 meses

from scipy.stats import zscore

# Estandarizar la serie original
zpre = zscore(pre)

# **BUSCAMOS EL MODELO CON MENOR $\text{AIC} $**

Se fue probando varias veces, y se buscaba ir encontrando el mejor $\text{AIC}$  en saltos, porque hacer todos los modelos tarda mucho tiempo. Por eso tiene mucho menos granulidad esta vez, que en la vez pasada.

In [61]:
def sarima(p, d, q, P, D, Q, s):    # La vamos a usar para calcular el AIC
    return SARIMAX(zpre, order=(p, d, q), seasonal_order=(P, D, Q, s)).fit(disp=0).aic    

Empezamos con el propuesto en 03

In [111]:
aics = []   # Lista para guardar los AICs

In [112]:
for P in range(0, 5):   # Para el componente estacional
    for Q in range(0, 5):   # Para el componente estacional
        aics.append((0, 0, 0, P, 1, Q, 12, sarima(0, 0, 0, P, 1, Q, 12)))

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date inde

In [113]:
sorted(aics, key=lambda x: x[7])

[(0, 0, 0, 0, 1, 3, 12, 1443.5046486930705),
 (0, 0, 0, 0, 1, 1, 12, 1443.5933641175748),
 (0, 0, 0, 2, 1, 1, 12, 1444.110845301806),
 (0, 0, 0, 1, 1, 2, 12, 1445.1467855295289),
 (0, 0, 0, 1, 1, 3, 12, 1445.2761261717749),
 (0, 0, 0, 0, 1, 4, 12, 1445.4876160065296),
 (0, 0, 0, 0, 1, 2, 12, 1445.5858681816667),
 (0, 0, 0, 1, 1, 1, 12, 1445.5874430840317),
 (0, 0, 0, 4, 1, 3, 12, 1445.703823120819),
 (0, 0, 0, 4, 1, 4, 12, 1445.7448183735353),
 (0, 0, 0, 2, 1, 2, 12, 1445.920682287981),
 (0, 0, 0, 4, 1, 2, 12, 1445.945951298871),
 (0, 0, 0, 3, 1, 1, 12, 1446.070360948445),
 (0, 0, 0, 2, 1, 3, 12, 1446.3837838663976),
 (0, 0, 0, 1, 1, 4, 12, 1446.56896129825),
 (0, 0, 0, 4, 1, 1, 12, 1446.6985652988965),
 (0, 0, 0, 3, 1, 2, 12, 1446.989869793793),
 (0, 0, 0, 3, 1, 3, 12, 1447.780088787008),
 (0, 0, 0, 2, 1, 4, 12, 1447.7999875463356),
 (0, 0, 0, 4, 1, 0, 12, 1448.2366123305535),
 (0, 0, 0, 3, 1, 4, 12, 1449.2145709622591),
 (0, 0, 0, 3, 1, 0, 12, 1454.1287421213299),
 (0, 0, 0, 2, 1, 0,

El modelo con menor $\text{AIC}$ es el $\text{ARIMA}(0,0,0) \times (0,1,3)$,  Con un $\text{AIC} = 1443.5046486930705$.

In [114]:
# Guardar los valores AIC en un archivo CSV
aic_df = pd.DataFrame(aics, columns=['p', 'd', 'q', 'P', 'D', 'Q', 's', 'AIC'])
aic_df.to_csv('aic_values-4.csv', index=False)

# **VISUALIZACIÓN**

In [115]:
import plotly.graph_objects as go
import numpy as np
from scipy.interpolate import griddata

In [116]:
# Separar en listas
p_values = [item[3] for item in aics]
q_values = [item[5] for item in aics]
aic_values = [item[7] for item in aics]

# Crear una malla regular
p_lin = np.linspace(min(p_values), max(p_values), 50)
q_lin = np.linspace(min(q_values), max(q_values), 50)
P, Q = np.meshgrid(p_lin, q_lin)

# Interpolar AIC en la malla
AIC_grid = griddata(
    points=(p_values, q_values),
    values=aic_values,
    xi=(P, Q),
    method='nearest'
)

In [117]:
# Crear figura 3D en Plotly
fig = go.Figure()

# Superficie
fig.add_trace(go.Surface(
    x=P, y=Q, z=AIC_grid,
    colorscale='Blues_r',
    opacity=0.7,
    colorbar=dict(title='AIC')
))

# Puntos originales
fig.add_trace(go.Scatter3d(
    x=p_values,
    y=q_values,
    z=aic_values,
    mode='markers',
    marker=dict(size=5, color=aic_values, colorscale='Blues_r'),
    name='Puntos'
))

# Layout
fig.update_layout(
    scene=dict(
        xaxis_title='P',
        yaxis_title='Q',
        zaxis_title='AIC'
    ),
    title='AIC para diferentes combinaciones de ARIMA(0,0,0)X(P,1,Q,12)',
    width=900,
    height=700
)

# Guardar como HTML
fig.write_html("AIC_3D_plot-4.html")
fig.show()

# **VERIFICACIÓN DE SUPUESTOS**

In [119]:
from statsmodels.tsa.statespace.sarimax import SARIMAX
from verificacion_de_supuestos import * # Una funcion que hice para verificar si los parametros son estacionarios/invertibles
from statsmodels.stats.diagnostic import acorr_ljungbox
from scipy.stats import ttest_1samp
from statsmodels.stats.diagnostic import het_breuschpagan
from statsmodels.tools.tools import add_constant
from scipy.stats import jarque_bera
from statsmodels.stats.diagnostic import lilliefors

## **MODELO 1**

In [120]:
modelo=SARIMAX(zpre,
               order=(0,0,0),
               seasonal_order=(0,1,3,12)).fit()

modelo.summary()

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.



<class 'statsmodels.iolib.summary.Summary'>
"""
                                    SARIMAX Results                                     
========================================================================================
Dep. Variable:                                y   No. Observations:                  660
Model:             SARIMAX(0, 1, [1, 2, 3], 12)   Log Likelihood                -717.752
Date:                          Sat, 26 Apr 2025   AIC                           1443.505
Time:                                  22:14:19   BIC                           1461.400
Sample:                                       0   HQIC                          1450.447
                                          - 660                                         
Covariance Type:                            opg                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
ma.S.L12      -0.6576      0.029    -22.456      0.000      -0.715      -0.600
ma.S.L24      -0.0539      0.037     -1.461      0.144      -0.126       0.018
ma.S.L36       0.0890      0.034      2.655      0.008       0.023       0.155
sigma2         0.5308      0.019     27.564      0.000       0.493       0.569
===================================================================================
Ljung-Box (L1) (Q):                  14.53   Jarque-Bera (JB):               260.95
Prob(Q):                              0.00   Prob(JB):                         0.00
Heteroskedasticity (H):               0.84   Skew:                             0.54
Prob(H) (two-sided):                  0.20   Kurtosis:                         5.92
===================================================================================

Warnings:
[1] Covariance matrix calculated using the outer product of gradients (complex-step).
"""

In [121]:
estacionario(modelo)

El polinomio de la parte autoregresiva es: {}

El grado del polinomio es: 0

Raíces del polinomio característico: []

Módulo de las raíces: []

¿Las raíces están fuera del círculo unitario?  True

El modelo es estacionario

:)


In [122]:
invertible(modelo)

El polinomio de la parte media móvil es: {'p12': -0.6576468674950458, 'p24': -0.05390724827273551, 'p36': 0.08902758389099319}

El grado del polinomio es: 36

Raíces del polinomio característico: [-1.05969906+0.28394551j -1.05969906-0.28394551j -1.0555843 +0.02843393j
 -1.0555843 -0.02843393j -0.92837979+0.50316764j -0.92837979-0.50316764j
 -0.89994585+0.55241666j -0.89994585-0.55241666j -0.77575355+0.77575355j
 -0.77575355-0.77575355j -0.55241666+0.89994585j -0.55241666-0.89994585j
 -0.50316764+0.92837979j -0.50316764-0.92837979j -0.28394551+1.05969906j
 -0.28394551-1.05969906j -0.02843393+1.0555843j  -0.02843393-1.0555843j
  0.02843393+1.0555843j   0.02843393-1.0555843j   0.28394551+1.05969906j
  0.28394551-1.05969906j  0.50316764+0.92837979j  0.50316764-0.92837979j
  0.55241666+0.89994585j  0.55241666-0.89994585j  0.77575355+0.77575355j
  0.77575355-0.77575355j  1.05969906+0.28394551j  1.05969906-0.28394551j
  1.0555843 +0.02843393j  1.0555843 -0.02843393j  0.89994585+0.55241666j
  

In [123]:
acorr_ljungbox(modelo.resid, period=24, model_df=len(modelo.params)-1)

lb_stat     lb_pvalue
1    17.179415           NaN
2    18.181363           NaN
3    18.881778           NaN
4    25.921220  3.556381e-07
5    51.167881  7.745256e-12
6    70.743004  2.959018e-15
7    72.370244  7.167546e-15
8    73.200464  2.207929e-14
9    74.909882  4.005840e-14
10   80.168939  1.272448e-14
11   92.113772  1.730068e-16
12   92.124138  6.096454e-16
13   92.141334  2.012622e-15
14   92.232794  6.080160e-15
15   92.362177  1.719885e-14
16   93.205534  3.395635e-14
17  102.044806  1.921292e-15
18  105.727402  1.060976e-15
19  106.843381  1.780329e-15
20  107.211226  4.029574e-15
21  107.213516  1.036857e-14
22  108.436542  1.546346e-14
23  111.753059  9.397567e-15
24  111.940952  2.101704e-14
25  113.076206  3.105212e-14
26  113.139589  7.011206e-14
27  114.208634  1.030170e-13
28  116.907961  7.742645e-14
29  119.404574  6.287217e-14
30  122.040366  4.807476e-14
31  122.883582  7.462406e-14
32  125.078484  6.721075e-14
33  125.367488  1.270263e-13
34  125.479975  2.532217e-13
35  127.469287  2.418063e-13
36  127.650123  4.585791e-13
37  128.428442  6.842664e-13
38  128.505727  1.314720e-12
39  128.507622  2.561399e-12
40  128.554185  4.840115e-12
41  128.747028  8.557366e-12
42  129.163273  1.379522e-11
43  129.208082  2.512976e-11
44  129.289079  4.464112e-11
45  129.294624  8.046093e-11
46  130.623145  9.070647e-11
47  131.535252  1.175387e-10
48  132.693491  1.392359e-10

Hay dependencia en los residuos.

## **MODELO 2**

In [129]:
modelo=SARIMAX(zpre,
               order=(0,0,0),
               seasonal_order=(0,1,1,12)).fit()

modelo.summary()

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.



<class 'statsmodels.iolib.summary.Summary'>
"""
                                 SARIMAX Results                                  
==================================================================================
Dep. Variable:                          y   No. Observations:                  660
Model:             SARIMAX(0, 1, [1], 12)   Log Likelihood                -719.797
Date:                    Sat, 26 Apr 2025   AIC                           1443.593
Time:                            22:14:20   BIC                           1452.541
Sample:                                 0   HQIC                          1447.064
                                    - 660                                         
Covariance Type:                      opg                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
ma.S.L12      -0.6533      0.025    -26.057      0.000      -0.702      -0.604
sigma2         0.5344      0.019     28.147      0.000       0.497       0.572
===================================================================================
Ljung-Box (L1) (Q):                  15.79   Jarque-Bera (JB):               286.71
Prob(Q):                              0.00   Prob(JB):                         0.00
Heteroskedasticity (H):               0.83   Skew:                             0.59
Prob(H) (two-sided):                  0.17   Kurtosis:                         6.04
===================================================================================

Warnings:
[1] Covariance matrix calculated using the outer product of gradients (complex-step).
"""

In [130]:
estacionario(modelo)

El polinomio de la parte autoregresiva es: {}

El grado del polinomio es: 0

Raíces del polinomio característico: []

Módulo de las raíces: []

¿Las raíces están fuera del círculo unitario?  True

El modelo es estacionario

:)


In [131]:
invertible(modelo)

El polinomio de la parte media móvil es: {'p12': -0.6532809672286413}

El grado del polinomio es: 12

Raíces del polinomio característico: [-1.03611589e+00+0.j         -8.97302679e-01+0.51805794j
 -8.97302679e-01-0.51805794j -5.18057943e-01+0.89730268j
 -5.18057943e-01-0.89730268j  8.48820604e-17+1.03611589j
  8.48820604e-17-1.03611589j  5.18057943e-01+0.89730268j
  5.18057943e-01-0.89730268j  1.03611589e+00+0.j
  8.97302679e-01+0.51805794j  8.97302679e-01-0.51805794j]

Módulo de las raíces: [1.03611589 1.03611589 1.03611589 1.03611589 1.03611589 1.03611589
 1.03611589 1.03611589 1.03611589 1.03611589 1.03611589 1.03611589]

¿Las raíces están fuera del círculo unitario?  True

El modelo es invertible

:)


In [132]:
acorr_ljungbox(modelo.resid, period=24, model_df=len(modelo.params)-1)

lb_stat     lb_pvalue
1    18.383847           NaN
2    19.527333  9.917048e-06
3    20.310714  3.886730e-05
4    27.528968  4.560262e-06
5    53.592080  6.406348e-11
6    73.402658  2.003677e-14
7    74.883915  4.055452e-14
8    75.550613  1.108538e-13
9    77.253111  1.743214e-13
10   83.190537  3.746241e-14
11   96.579973  2.629359e-16
12   96.580852  8.467666e-16
13   96.602684  2.570188e-15
14   96.644007  7.393307e-15
15   96.961689  1.807724e-14
16   98.060502  3.038103e-14
17  108.019725  1.065783e-15
18  111.782010  5.571411e-16
19  112.957798  8.831177e-16
20  113.234782  2.013579e-15
21  113.236528  5.027918e-15
22  114.683673  6.681108e-15
23  118.206110  3.689699e-15
24  118.935786  6.462842e-15
25  120.087456  9.361005e-15
26  120.144340  2.088241e-14
27  121.296370  2.937399e-14
28  124.360591  1.898206e-14
29  126.797911  1.572230e-14
30  129.784562  1.042222e-14
31  131.242512  1.262727e-14
32  133.790099  9.888833e-15
33  134.252478  1.745355e-14
34  134.378908  3.463331e-14
35  136.331235  3.359133e-14
36  136.959944  5.381216e-14
37  137.884613  7.625907e-14
38  137.966595  1.471026e-13
39  137.966983  2.886599e-13
40  138.136806  5.250562e-13
41  138.458291  8.928893e-13
42  139.145348  1.316054e-12
43  139.221849  2.396342e-12
44  139.351337  4.232646e-12
45  139.360719  7.714488e-12
46  141.058592  7.702387e-12
47  142.573638  8.175892e-12
48  142.765655  1.367752e-11

Hay dependencia en los residuos.

## **MODELO 3**

In [138]:
modelo=SARIMAX(zpre,
               order=(0,0,0),
               seasonal_order=(2,1,1,12)).fit()

modelo.summary()

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.



<class 'statsmodels.iolib.summary.Summary'>
"""
                                 SARIMAX Results                                  
==================================================================================
Dep. Variable:                          y   No. Observations:                  660
Model:             SARIMAX(2, 1, [1], 12)   Log Likelihood                -718.055
Date:                    Sat, 26 Apr 2025   AIC                           1444.111
Time:                            22:14:20   BIC                           1462.006
Sample:                                 0   HQIC                          1451.053
                                    - 660                                         
Covariance Type:                      opg                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
ar.S.L12      -0.1041      0.069     -1.501      0.133      -0.240       0.032
ar.S.L24      -0.1141      0.050     -2.277      0.023      -0.212      -0.016
ma.S.L12      -0.5526      0.072     -7.689      0.000      -0.693      -0.412
sigma2         0.5313      0.019     27.725      0.000       0.494       0.569
===================================================================================
Ljung-Box (L1) (Q):                  14.85   Jarque-Bera (JB):               267.24
Prob(Q):                              0.00   Prob(JB):                         0.00
Heteroskedasticity (H):               0.84   Skew:                             0.55
Prob(H) (two-sided):                  0.19   Kurtosis:                         5.95
===================================================================================

Warnings:
[1] Covariance matrix calculated using the outer product of gradients (complex-step).
"""

In [139]:
estacionario(modelo)

El polinomio de la parte autoregresiva es: {'p12': 0.10410933674538352, 'p24': 0.11410119085946407}

El grado del polinomio es: 24

Raíces del polinomio característico: [-1.08336346+0.1568628j  -1.08336346-0.1568628j  -1.01665168+0.40583456j
 -1.01665168-0.40583456j -0.85978888+0.6775289j  -0.85978888-0.6775289j
 -0.6775289 +0.85978888j -0.6775289 -0.85978888j -0.40583456+1.01665168j
 -0.40583456-1.01665168j -0.1568628 +1.08336346j -0.1568628 -1.08336346j
  0.1568628 +1.08336346j  0.1568628 -1.08336346j  0.40583456+1.01665168j
  0.40583456-1.01665168j  0.6775289 +0.85978888j  0.6775289 -0.85978888j
  0.85978888+0.6775289j   0.85978888-0.6775289j   1.08336346+0.1568628j
  1.08336346-0.1568628j   1.01665168+0.40583456j  1.01665168-0.40583456j]

Módulo de las raíces: [1.09466082 1.09466082 1.09466082 1.09466082 1.09466082 1.09466082
 1.09466082 1.09466082 1.09466082 1.09466082 1.09466082 1.09466082
 1.09466082 1.09466082 1.09466082 1.09466082 1.09466082 1.09466082
 1.09466082 1.09466082 1

In [140]:
invertible(modelo)

El polinomio de la parte media móvil es: {'p12': -0.5526053974957342}

El grado del polinomio es: 12

Raíces del polinomio característico: [-1.05066776e+00+0.j         -9.09904972e-01+0.52533388j
 -9.09904972e-01-0.52533388j -5.25333881e-01+0.90990497j
 -5.25333881e-01-0.90990497j -2.54511798e-16+1.05066776j
 -2.54511798e-16-1.05066776j  5.25333881e-01+0.90990497j
  5.25333881e-01-0.90990497j  1.05066776e+00+0.j
  9.09904972e-01+0.52533388j  9.09904972e-01-0.52533388j]

Módulo de las raíces: [1.05066776 1.05066776 1.05066776 1.05066776 1.05066776 1.05066776
 1.05066776 1.05066776 1.05066776 1.05066776 1.05066776 1.05066776]

¿Las raíces están fuera del círculo unitario?  True

El modelo es invertible

:)


In [141]:
acorr_ljungbox(modelo.resid, period=24, model_df=len(modelo.params)-1)

lb_stat     lb_pvalue
1    17.493428           NaN
2    18.519956           NaN
3    19.228301           NaN
4    26.315332  2.899775e-07
5    51.914435  5.332412e-12
6    71.654631  1.887537e-15
7    73.318872  4.517343e-15
8    74.131900  1.411701e-14
9    75.884632  2.523257e-14
10   81.283735  7.536739e-15
11   93.596985  8.636585e-17
12   93.608257  3.065902e-16
13   93.624656  1.020430e-15
14   93.709905  3.115383e-15
15   93.858972  8.802065e-15
16   94.765772  1.701633e-14
17  103.885049  8.503266e-16
18  107.608770  4.633906e-16
19  108.733860  7.802304e-16
20  109.071815  1.803565e-15
21  109.073437  4.681252e-15
22  110.365561  6.826147e-15
23  113.845059  3.888463e-15
24  113.903689  9.260346e-15
25  115.063033  1.363864e-14
26  115.123033  3.109739e-14
27  116.222202  4.544185e-14
28  119.072369  3.226096e-14
29  121.632795  2.564847e-14
30  124.340372  1.913674e-14
31  125.241688  2.924512e-14
32  127.434555  2.649330e-14
33  127.737882  5.021757e-14
34  127.881541  9.978540e-14
35  129.957766  9.260264e-14
36  130.086849  1.807596e-13
37  130.928539  2.652767e-13
38  131.002266  5.150984e-13
39  131.006397  1.012139e-12
40  131.078945  1.911939e-12
41  131.305227  3.369331e-12
42  131.781692  5.359562e-12
43  131.841413  9.805024e-12
44  131.933690  1.751507e-11
45  131.939852  3.187362e-11
46  133.402913  3.448353e-11
47  134.427250  4.329434e-11
48  135.426547  5.453836e-11

Hay dependencia en los residuos.

## **MODELO 4**

In [147]:
modelo=SARIMAX(zpre,
               order=(0,0,0),
               seasonal_order=(1,1,2,12)).fit()

modelo.summary()

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.



<class 'statsmodels.iolib.summary.Summary'>
"""
                                   SARIMAX Results                                   
=====================================================================================
Dep. Variable:                             y   No. Observations:                  660
Model:             SARIMAX(1, 1, [1, 2], 12)   Log Likelihood                -718.573
Date:                       Sat, 26 Apr 2025   AIC                           1445.147
Time:                               22:14:22   BIC                           1463.042
Sample:                                    0   HQIC                          1452.089
                                       - 660                                         
Covariance Type:                         opg                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
ar.S.L12      -0.8935      0.138     -6.496      0.000      -1.163      -0.624
ma.S.L12       0.2628      0.132      1.985      0.047       0.003       0.522
ma.S.L24      -0.6117      0.077     -7.912      0.000      -0.763      -0.460
sigma2         0.5321      0.019     28.185      0.000       0.495       0.569
===================================================================================
Ljung-Box (L1) (Q):                  14.91   Jarque-Bera (JB):               285.36
Prob(Q):                              0.00   Prob(JB):                         0.00
Heteroskedasticity (H):               0.84   Skew:                             0.57
Prob(H) (two-sided):                  0.20   Kurtosis:                         6.05
===================================================================================

Warnings:
[1] Covariance matrix calculated using the outer product of gradients (complex-step).
"""

In [148]:
estacionario(modelo)

El polinomio de la parte autoregresiva es: {'p12': 0.8934837600190328}

El grado del polinomio es: 12

Raíces del polinomio característico: [-0.97503429+0.26125965j -0.97503429-0.26125965j -0.71377464+0.71377464j
 -0.71377464-0.71377464j -0.26125965+0.97503429j -0.26125965-0.97503429j
  0.26125965+0.97503429j  0.26125965-0.97503429j  0.71377464+0.71377464j
  0.71377464-0.71377464j  0.97503429+0.26125965j  0.97503429-0.26125965j]

Módulo de las raíces: [1.00942978 1.00942978 1.00942978 1.00942978 1.00942978 1.00942978
 1.00942978 1.00942978 1.00942978 1.00942978 1.00942978 1.00942978]

¿Las raíces están fuera del círculo unitario?  True

El modelo es estacionario

:)


In [149]:
invertible(modelo)

El polinomio de la parte media móvil es: {'p12': 0.2628444255001554, 'p24': -0.6117107674153909}

El grado del polinomio es: 24

Raíces del polinomio característico: [-1.03501581e+00+0.j         -9.72264921e-01+0.2605176j
 -9.72264921e-01-0.2605176j  -8.96349985e-01+0.51750791j
 -8.96349985e-01-0.51750791j -7.11747321e-01+0.71174732j
 -7.11747321e-01-0.71174732j -5.17507905e-01+0.89634999j
 -5.17507905e-01-0.89634999j -2.60517601e-01+0.97226492j
 -2.60517601e-01-0.97226492j -4.10915192e-16+1.03501581j
 -4.10915192e-16-1.03501581j  2.60517601e-01+0.97226492j
  2.60517601e-01-0.97226492j  5.17507905e-01+0.89634999j
  5.17507905e-01-0.89634999j  7.11747321e-01+0.71174732j
  7.11747321e-01-0.71174732j  1.03501581e+00+0.j
  9.72264921e-01+0.2605176j   9.72264921e-01-0.2605176j
  8.96349985e-01+0.51750791j  8.96349985e-01-0.51750791j]

Módulo de las raíces: [1.03501581 1.00656271 1.00656271 1.03501581 1.03501581 1.00656271
 1.00656271 1.03501581 1.03501581 1.00656271 1.00656271 1.03501581
 1

In [150]:
acorr_ljungbox(modelo.resid, period=24, model_df=len(modelo.params)-1)

lb_stat     lb_pvalue
1    17.463603           NaN
2    18.645709           NaN
3    19.472263           NaN
4    26.854810  2.193278e-07
5    52.510958  3.957221e-12
6    72.359787  1.333039e-15
7    74.034847  3.188012e-15
8    74.653056  1.099051e-14
9    76.187459  2.185605e-14
10   82.058149  5.236751e-15
11   96.146093  2.612564e-17
12   96.300868  8.790949e-17
13   96.402285  2.853016e-16
14   96.456439  8.960154e-16
15   96.813119  2.338249e-15
16   97.869227  4.287590e-15
17  107.704972  1.556966e-16
18  111.542715  8.148025e-17
19  112.768897  1.332318e-16
20  113.134189  3.095563e-16
21  113.135200  8.181654e-16
22  114.606282  1.121291e-15
23  117.806556  7.254977e-16
24  117.823739  1.787518e-15
25  118.803765  2.875346e-15
26  118.856746  6.677508e-15
27  119.904335  1.009110e-14
28  123.130510  6.188529e-15
29  125.716447  4.910318e-15
30  128.830624  3.131761e-15
31  130.328914  3.819035e-15
32  132.561235  3.442305e-15
33  132.954498  6.410762e-15
34  133.111559  1.291382e-14
35  135.558774  1.047507e-14
36  135.682840  2.090070e-14
37  136.845293  2.755979e-14
38  136.928728  5.445721e-14
39  136.929283  1.094848e-13
40  137.098794  2.037023e-13
41  137.482177  3.456666e-13
42  138.232207  5.069118e-13
43  138.326159  9.370453e-13
44  138.399198  1.724367e-12
45  138.399246  3.218764e-12
46  140.132066  3.205271e-12
47  141.480487  3.647829e-12
48  141.485056  6.632294e-12

Hay dependencia en los residuos.

## **MODELO 5**

In [156]:
modelo=SARIMAX(zpre,
               order=(1,0,3),
               seasonal_order=(0,1,4,12)).fit()

modelo.summary()

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.



<class 'statsmodels.iolib.summary.Summary'>
"""
                                           SARIMAX Results                                           
=====================================================================================================
Dep. Variable:                                             y   No. Observations:                  660
Model:             SARIMAX(1, 0, 3)x(0, 1, [1, 2, 3, 4], 12)   Log Likelihood                -707.513
Date:                                       Sat, 26 Apr 2025   AIC                           1433.026
Time:                                               22:14:51   BIC                           1473.291
Sample:                                                    0   HQIC                          1448.646
                                                       - 660                                         
Covariance Type:                                         opg                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
ar.L1         -0.1204      0.438     -0.275      0.783      -0.978       0.738
ma.L1          0.2845      0.436      0.652      0.514      -0.571       1.140
ma.L2          0.0874      0.080      1.090      0.276      -0.070       0.245
ma.L3          0.1134      0.044      2.548      0.011       0.026       0.201
ma.S.L12      -0.6893      0.029    -23.542      0.000      -0.747      -0.632
ma.S.L24      -0.0505      0.037     -1.366      0.172      -0.123       0.022
ma.S.L36       0.0709      0.038      1.846      0.065      -0.004       0.146
ma.S.L48      -0.0032      0.034     -0.093      0.926      -0.070       0.064
sigma2         0.5135      0.020     25.670      0.000       0.474       0.553
===================================================================================
Ljung-Box (L1) (Q):                   0.00   Jarque-Bera (JB):               207.43
Prob(Q):                              0.98   Prob(JB):                         0.00
Heteroskedasticity (H):               0.86   Skew:                             0.44
Prob(H) (two-sided):                  0.25   Kurtosis:                         5.63
===================================================================================

Warnings:
[1] Covariance matrix calculated using the outer product of gradients (complex-step).
"""

In [157]:
estacionario(modelo)

El polinomio de la parte autoregresiva es: {'p1': 0.12042050604759126}

El grado del polinomio es: 1

Raíces del polinomio característico: [-8.3042335]

Módulo de las raíces: [8.3042335]

¿Las raíces están fuera del círculo unitario?  True

El modelo es estacionario

:)


In [158]:
invertible(modelo)

El polinomio de la parte media móvil es: {'p1': 0.28453739913448367, 'p2': 0.08744263796213617, 'p3': 0.11336638804641153, 'p12': -0.689349750524581, 'p24': -0.050470186215013375, 'p36': 0.0708803762819837, 'p48': -0.0031726073065468717}

El grado del polinomio es: 48

Raíces del polinomio característico: [ 1.28962843e+00+0.j          1.11674298e+00+0.64486861j
  1.11674298e+00-0.64486861j  1.08048252e+00+0.04326004j
  1.08048252e+00-0.04326004j  1.06965373e+00+0.28919708j
  1.06965373e+00-0.28919708j  6.44662256e-01+1.11675118j
  6.44662256e-01-1.11675118j  9.63397802e-01+0.52088179j
  9.63397802e-01-0.52088179j  8.91652798e-01+0.57127551j
  8.91652798e-01-0.57127551j  7.77061882e-01+0.78436918j
  7.77061882e-01-0.78436918j -2.28099765e-05+1.28946341j
 -2.28099765e-05-1.28946341j  5.64659117e-01+0.94385916j
  5.64659117e-01-0.94385916j  4.83589283e-01+0.91585203j
  4.83589283e-01-0.91585203j  2.81736324e-01+1.0642343j
  2.81736324e-01-1.0642343j   8.52825570e-03+1.09648617j
  8.528255

In [159]:
acorr_ljungbox(modelo.resid, period=24, model_df=len(modelo.params)-1)

lb_stat     lb_pvalue
1    0.127577           NaN
2    0.232730           NaN
3    0.773182           NaN
4    2.709120           NaN
5   20.690359           NaN
6   35.530901           NaN
7   35.928114           NaN
8   36.837023           NaN
9   38.638278  5.100869e-10
10  41.799973  8.380137e-10
11  53.352772  1.541821e-11
12  53.362886  7.154599e-11
13  53.363070  2.832799e-10
14  53.509370  9.263900e-10
15  53.510511  2.938971e-09
16  53.783665  7.603558e-09
17  62.359657  4.694183e-10
18  64.984049  4.081871e-10
19  65.575335  8.398309e-10
20  65.695505  2.027131e-09
21  65.703170  4.911513e-09
22  66.709203  7.565006e-09
23  70.360228  3.853223e-09
24  70.432975  8.379152e-09
25  71.755187  1.073183e-08
26  71.774112  2.258796e-08
27  72.394110  3.662448e-08
28  74.713397  3.040967e-08
29  75.927845  3.824625e-08
30  78.016445  3.424973e-08
31  78.283368  6.008154e-08
32  80.795965  4.541200e-08
33  81.702054  6.159194e-08
34  81.720283  1.135628e-07
35  84.657873  7.295312e-08
36  84.749345  1.282915e-07
37  85.816606  1.580931e-07
38  85.822140  2.782519e-07
39  85.825411  4.818814e-07
40  85.942834  7.905316e-07
41  86.270166  1.193548e-06
42  86.955364  1.586976e-06
43  87.111514  2.475290e-06
44  87.155467  3.947454e-06
45  87.161504  6.283078e-06
46  88.536969  6.522862e-06
47  89.914721  6.750239e-06
48  90.882955  7.875131e-06

Hay dependencia en los residuos.

## **MODELO 6**

In [165]:
modelo=SARIMAX(zpre,
               order=(0,0,0),
               seasonal_order=(0,1,2,12)).fit()

modelo.summary()

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.

c:\ProgramData\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning:

A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.



<class 'statsmodels.iolib.summary.Summary'>
"""
                                   SARIMAX Results                                   
=====================================================================================
Dep. Variable:                             y   No. Observations:                  660
Model:             SARIMAX(0, 1, [1, 2], 12)   Log Likelihood                -719.793
Date:                       Sat, 26 Apr 2025   AIC                           1445.586
Time:                               22:14:52   BIC                           1459.008
Sample:                                    0   HQIC                          1450.793
                                       - 660                                         
Covariance Type:                         opg                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
ma.S.L12      -0.6560      0.029    -22.764      0.000      -0.712      -0.599
ma.S.L24       0.0038      0.031      0.122      0.903      -0.057       0.064
sigma2         0.5344      0.019     27.963      0.000       0.497       0.572
===================================================================================
Ljung-Box (L1) (Q):                  15.75   Jarque-Bera (JB):               285.46
Prob(Q):                              0.00   Prob(JB):                         0.00
Heteroskedasticity (H):               0.83   Skew:                             0.59
Prob(H) (two-sided):                  0.18   Kurtosis:                         6.03
===================================================================================

Warnings:
[1] Covariance matrix calculated using the outer product of gradients (complex-step).
"""

In [166]:
estacionario(modelo)

El polinomio de la parte autoregresiva es: {}

El grado del polinomio es: 0

Raíces del polinomio característico: []

Módulo de las raíces: []

¿Las raíces están fuera del círculo unitario?  True

El modelo es estacionario

:)


In [167]:
invertible(modelo)

El polinomio de la parte media móvil es: {'p12': -0.6559709015852816, 'p24': 0.0037660425032464548}

El grado del polinomio es: 24

Raíces del polinomio característico: [-1.53613348e+00+0.j         -1.33033061e+00+0.76806674j
 -1.33033061e+00-0.76806674j -7.68066738e-01+1.33033061j
 -7.68066738e-01-1.33033061j -9.85778834e-16+1.53613348j
 -9.85778834e-16-1.53613348j  7.68066738e-01+1.33033061j
  7.68066738e-01-1.33033061j  1.33033061e+00+0.76806674j
  1.33033061e+00-0.76806674j  1.53613348e+00+0.j
 -1.03652698e+00+0.j         -8.97658698e-01+0.51826349j
 -8.97658698e-01-0.51826349j -5.18263491e-01+0.8976587j
 -5.18263491e-01-0.8976587j  -1.05918390e-16+1.03652698j
 -1.05918390e-16-1.03652698j  5.18263491e-01+0.8976587j
  5.18263491e-01-0.8976587j   8.97658698e-01+0.51826349j
  8.97658698e-01-0.51826349j  1.03652698e+00+0.j        ]

Módulo de las raíces: [1.53613348 1.53613348 1.53613348 1.53613348 1.53613348 1.53613348
 1.53613348 1.53613348 1.53613348 1.53613348 1.53613348 1.53613348

In [168]:
acorr_ljungbox(modelo.resid, period=24, model_df=len(modelo.params)-1)

lb_stat     lb_pvalue
1    18.349617           NaN
2    19.494151           NaN
3    20.284211  6.674915e-06
4    27.484417  1.076056e-06
5    53.552157  1.398039e-11
6    73.369953  4.406415e-15
7    74.849287  1.000165e-14
8    75.515860  3.005519e-14
9    77.220547  5.072682e-14
10   83.141255  1.137665e-14
11   96.539389  7.869032e-17
12   96.547423  2.668977e-16
13   96.568150  8.516686e-16
14   96.607951  2.564111e-15
15   96.923805  6.528803e-15
16   98.027386  1.131277e-14
17  108.010054  3.882444e-16
18  111.765597  2.069343e-16
19  112.918400  3.400168e-16
20  113.196222  7.969425e-16
21  113.197815  2.045571e-15
22  114.639768  2.778816e-15
23  118.156724  1.553684e-15
24  118.963303  2.690062e-15
25  120.089683  4.008972e-15
26  120.146392  9.137410e-15
27  121.294676  1.308132e-14
28  124.336180  8.597720e-15
29  126.754232  7.246073e-15
30  129.705349  4.906580e-15
31  131.145246  6.060540e-15
32  133.687148  4.793175e-15
33  134.139408  8.618163e-15
34  134.262391  1.738615e-14
35  136.182434  1.721733e-14
36  136.756541  2.851820e-14
37  137.678361  4.089619e-14
38  137.761234  7.990888e-14
39  137.761546  1.588735e-13
40  137.925098  2.931921e-13
41  138.231591  5.070269e-13
42  138.901939  7.592068e-13
43  138.974954  1.399891e-12
44  139.104497  2.499465e-12
45  139.115377  4.603127e-12
46  140.794967  4.649888e-12
47  142.294787  4.989972e-12
48  142.501464  8.384217e-12

Hay dependencia en los residuos.